examples/src/bin/gio_task.rs

```diff
@@ -0,0 +1,188 @@
+use gtk::{gio, glib};
+use std::convert::TryInto;
+
+pub mod imp {
+    use super::*;
+
+    // FileSize is a simple object that will just contain the read file size.
+    // Initially the size field will be initialized to 0.
+    pub type FileSize = <FileSizePrivate as ObjectSubclass>::Instance;
+    pub type FileSizeClass = <FileSizePrivate as ObjectSubclass>::Class;
+    pub struct FileSizePrivate {
+        pub size: std::cell::RefCell<i64>,
+    }
+
+    impl ObjectSubclass for FileSizePrivate {
+        const NAME: &'static str = "FileSize";
+        type ParentType = glib::Object;
+        type Instance = subclass::simple::InstanceStruct<Self>;
+        type Class = subclass::simple::ClassStruct<Self>;
+        type Type = super::FileSize;
+        glib::object_subclass!();
+
+        fn new() -> Self {
+            Self {
+                size: std::cell::RefCell::new(0),
+            }
+        }
+    }
+
+    impl ObjectImpl for FileSizePrivate {}
+
+    /// # Safety
+    ///
+    /// This is the ffi method to asynchronously get the file size. It accepts a callback of type
+    /// GAsyncReadyCallback, that will be invoked when the async operation completes. Typically,
+    /// this callback will invoke the get_file_size_finish method (implemented below) to get the
+    /// Task result and perform some operation with it.
+    #[no_mangle]
+    pub unsafe extern "C" fn get_file_size_async(
+        this: *mut FileSize,
+        cancellable: *mut gio::ffi::GCancellable,
+        callback: gio::ffi::GAsyncReadyCallback,
+        user_data: glib::ffi::gpointer,
+    ) {
+        let cancellable = gio::Cancellable::from_glib_none(cancellable);
+        let closure = move |result: &gio::AsyncResult, source_object: Option<&glib::Object>| {
+            let result: *mut gio::ffi::GAsyncResult = result.to_glib_none().0;
+            let source_object: *mut glib::object::GObject = source_object.to_glib_none().0;
+            callback.unwrap()(source_object, result, user_data)
+        };
+
+        let source_object = &super::FileSize::from_glib_none(this);
+        let task = gio::Task::new(
+            Some(&source_object.clone().upcast::<glib::Object>()),
+            Some(&cancellable),
+            closure,
+        );
+
+        glib::MainContext::default().spawn_local(async move {
+            let size = gio::File::new_for_path("Cargo.toml")
+                .query_info_async_future("*", gio::FileQueryInfoFlags::NONE, glib::PRIORITY_DEFAULT)
+                .await
+                .unwrap()
+                .get_size();
+
+            let source_object = task
+                .upcast_ref::<gio::AsyncResult>()
+                .get_source_object()
+                .unwrap();
+
+            let source_object = FileSizePrivate::from_instance(
+                &source_object.downcast_ref::<super::FileSize>().unwrap(),
+            );
+
+            source_object.size.replace(size);
+            task.return_int(size.try_into().unwrap());
```

Comment:
Maybe return as `glib::Value` so you can have a proper `i64` or `u64` instead of having to work with `isize` and failing on 32 bit platforms if the file is bigger than 2GB